File: codes/025.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @param {Integer} k
# @return {ListNode}
def reverse_k_group(head, k)
    return [] if head==nil
    arrs = []
    while head
      arrs << head.val
      head = head.next
    end


    arrs = arrs.each_slice(k).to_a
    if arrs[-1].length != k
        t = arrs.pop
    end
    arrs = arrs.map(&:reverse).flatten+Array(t)


  l = nil
  v = nil
  arrs.each_with_index do |item, index|
    if index == 0
      l = ListNode.new(item)
      v = l.next = ListNode.new(0) unless index+1 == arrs.length
    else
      v.val=item
      v = v.next = ListNode.new(0) unless index+1 == arrs.length
    end
  end
  l
end

File: codes/025.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
    if head==nil||k==1{return head}
    c:=&ListNode{Next: head}
    pre,cur,count:=c,head,0
    reverse:=func(p,n *ListNode)*ListNode{
        last := p.Next
        current := last.Next
        for current!=n{
            last.Next = current.Next
            current.Next = pre.Next
            pre.Next = current
            current = last.Next
        }
        return last
    }
    for cur!=nil{
        count++
        if count%k==0{
            pre = reverse(pre,cur.Next)
            cur = pre.Next
        }else{
            cur = cur.Next
        }
    }
    return c.Next
}
